classes/module/Module.php

```diff
@@ -3436,30 +3442,70 @@ public function isSymfonyContext()
     }
 
     /**
-     * Access the Symfony Container if we are in Symfony Context.
-     * Note: in this case, we must get a container from SymfonyContainer class.
-     * Note: if not in Symfony context, fallback to legacy Container for FO/BO.
+     * @return bool
+     */
+    public function hasSymfonyContainer(): bool
+    {
+        // Check that a container is available but not generated by PrestaShop
+        $container = $this->getContainer();
+
+        return null !== $container && !($container instanceof LegacyContainerInterface);
+    }
+
+    /**
+     * @return bool
+     */
+    public function hasLegacyContainer(): bool
+    {
+        // Check that a container is available but not generated by PrestaShop
+        $container = $this->getContainer();
+
+        return null !== $container && $container instanceof LegacyContainerInterface;
+    }
+
+    /**
+     * Access a service from the container (found from the controller or the context
+     * depending on cases). It uses ContainerFinder to find the appropriate container.
      *
      * @param string $serviceName
      *
-     * @return object|false if a container is not available, it returns false
+     * @return object|false If a container is not available it returns false
+     *
+     * @throws ServiceCircularReferenceException When a circular reference is detected
+     * @throws ServiceNotFoundException When the service is not defined
+     * @throws \Exception
      */
     public function get($serviceName)
     {
-        if ($this->isSymfonyContext()) {
-            if (null === $this->container) {
-                $this->container = SymfonyContainer::getInstance();
-            }
-
-            return $this->container->get($serviceName);
+        try {
+            $container = $this->getContainer();
+        } catch (ContainerNotFoundException $e) {
+            $container = null;
         }
-
-        if ($this->context->controller instanceof Controller) {
-            return $this->context->controller->get($serviceName);
+        if (null !== $container) {
+            return $container->get($serviceName);
         }
 
         return false;
     }
+
+    /**
+     * Return the container depending on the context:
+     *  - Symfony environment: container generated by the kernel
+     *  - Legacy environment admin: container generated by PrestaShop core for admin controllers
+     *  - Legacy environment front: container generated by PrestaShop core for front controllers
+     *
+     * @return ContainerBuilder|ContainerInterface|null
```

Comment:
I suppose all containers implement `ContainerInterface`? In that case you can just hint that interface. And add it as a return type.

Reply:
BTW this return type is also wrong in `ContainerFinder` which will never return null.

Reply:
I removed `ContainerBuilder` indeed, but I put `LegacyContainerInterface` instead to emphasize the fact that two king of containers can be returned (Symfony and legacy)

Reply:
They still belong to the same class hierarchy